noir-projects/noir-contracts/contracts/nft_contract/src/main.nr

```diff
@@ -134,21 +135,41 @@ contract NFT {
         public_owners_storage.write(to);
     }
 
-    /// Prepares a transfer from public balance of `from` to a private balance of `to`. The transfer then needs to be
-    /// finalized by calling `finalize_transfer_to_private`. `transient_storage_slot_randomness` is passed
-    /// as an argument so that we can derive `transfer_preparer_storage_slot_commitment` off-chain and then pass it
-    /// as an argument to the followup call to `finalize_transfer_to_private`.
+    // Transfers token with `token_id` from public balance of message sender to a private balance of `to`.
     #[private]
-    fn prepare_transfer_to_private(
-        from: AztecAddress,
-        to: AztecAddress,
-        transient_storage_slot_randomness: Field
-    ) {
+    fn transfer_to_private(to: AztecAddress, token_id: Field) {
+        let from = context.msg_sender();
+
+        let nft = NFT::at(context.this_address());
+
+        // We prepare the transfer.
+        let hiding_point_slot = nft._prepare_transfer_to_private(to).call(&mut context);
+
+        // At last we finalize the transfer. Usafe of the `unsafe` method here is safe because we set the `from`
+        // function argument to a message sender, guaranteeing that he can transfer only his own NFTs.
+        nft._finalize_transfer_to_private_unsafe(from, token_id, hiding_point_slot).enqueue(&mut context);
+    }
+
+    /// Prepares a transfer to a private balance of `to`. The transfer then needs to be
+    /// finalized by calling `finalize_transfer_to_private`. Returns a hiding point slot.
+    #[private]
+    fn prepare_transfer_to_private(to: AztecAddress) -> Field {
+        NFT::at(context.this_address())._prepare_transfer_to_private(to).call(&mut context)
+    }
+
+    /// This function exists separately from `prepare_transfer_to_private` solely as an optimization as it allows
+    /// us to have it inlined in the `transfer_to_private` function which results in one less kernel iteration.
+    ///
+    /// TODO(#9180): Consider adding macro support for functions callable both as an entrypoint and as an internal
+    /// function.
+    #[private]
+    #[internal]
+    fn _prepare_transfer_to_private(to: AztecAddress) -> Field {
         let to_keys = get_public_keys(to);
         let to_npk_m_hash = to_keys.npk_m.hash();
         let to_note_slot = storage.private_nfts.at(to).storage_slot;
 
-        // We create a partial NFT note hiding point with unpopulated/zero token id for 'to'
+        // We create a setup payload with unpopulated/zero token id for 'to'
         // TODO(#7775): Manually fetching the randomness here is not great. If we decide to include randomness in all
```

Comment:
We're already calling `random` inside the `new` function of other notes, see e.g. [here](https://github.com/AztecProtocol/aztec-packages/pull/8143/files#diff-0610e54fc25c4c0e203ba8914010d5d7ec3018b72f87bd28e6d48ee94ffd0430R52).

Reply:
I think it's just a stale code from when we needed the randomness exposed due to the old storage slot derivation scheme.

I will address this in a PR up the stack as I am touching the macros there.

Reply:
@nventuro I was being stupid. It's not a stale code but instead it's done this way because the setup payload is auto generated and hence I cannot easily modify the SetupPayload constructor here. Makes sense to tackle this when when dealing with [the linked issue](https://github.com/AztecProtocol/aztec-packages/issues/7775) as then we'll most likely mandatorily include the randomness in the note and then we can afford to hardcode it in all the `SetupPayload`s.